### PR TITLE
feat(frontend): RFC to incubate React app

### PR DESCRIPTION
**Pull request by jjoyce0510**
Rendered: https://github.com/linkedin/datahub/blob/762ce4a0dcf20ab8ad413f22373dcefcd1ceff3a/docs/rfc/active/react-app/README.md

~ This PR includes a proposal to introduce a React application to be incubated within DataHub ~

## Checklist
- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [x] Docs related to the changes have been added/updated (if applicable)

### CODE REVIEW

**Comment by shirshanka** [comment]
Agree with the general direction towards React(w/GraphQL) and proposal to incubate versus boil the ocean.
We have discussed this within LinkedIn several times as well.

File: docs/rfc/active/react-app/README.md [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+- Managing a user account
+
+The finer details of which entities fall into each feature bucket will be dictated by the needs of the community, with the short-term milestone to achieve parity with entities appearing in the Ember client (Datasets, CorpUsers).
+
+
+#### Long term
+
+In the longer term, we will work with the community to define a more extensive functional road map, which may include
+
+- Providing migration pathway from the Ember application to the React application
```
Comment (jplaisted) [inline docs/rfc/active/react-app/README.md:98]: I would ask that this be done before any code commits. Ditto with GQL, tbh.

Until we have migration plans we'd need to maintain the old world too. So if we have two options for the UI (react, ember) and two for the midtier (GQL, legacy restli), then suddenly we also have 4 combinations of things to support, and the complexity of the software greatly increases.

Thinking about EoL for the things you want to replace should be done ahead of time. You might find it isn't as straightforward or feasible as you thought, and imo that's a pretty big cost to adding the new thing as a result.

Comment (jplaisted) [reply docs/rfc/active/react-app/README.md:98]: I guess if we also say "ember only supports legacy midtier; react only GQL" then its only two states, but might make it more difficult to get a deployment correct.

Comment (jplaisted) [reply docs/rfc/active/react-app/README.md:98]: Hmm maybe this isn't as big of a deal if the stacks are separate, you just flip a switch at some point and start serving the other app I guess. I think the only concern is getting apollo deployed, but that should be straight forward?

I guess I'm more ok with this being delayed if we go with this approach. Resolving. lol.

Comment (shirshanka) [reply docs/rfc/active/react-app/README.md:98]: Yes my understanding of this proposal is that the stacks are separate as you describe it. @jjoyce0510 to confirm.

Comment (jplaisted) [reply docs/rfc/active/react-app/README.md:98]: Yep, its later down in the RFC. Which is why I resolved this as I read more :p

Comment (jjoyce0510) [reply docs/rfc/active/react-app/README.md:98]: Your understanding is correct! 

"Getting apollo deployed" --> Not following this piece. If you mean getting a GraphQL endpoint at the frontend layer deployed, then yes it's true that will need to exist

File: docs/rfc/active/react-app/README.md [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+*What?*: Iterate on the existing Ember appclient.
+
+*Why not?* Firstly, we actually do not consider this to be mutually exclusive with introducing a separate React app. Regardless, there are benefits to adopting a more accessible technology like React that do not change with improvements to the existing Ember app.
+
+### Mixing Ember & React
+
+*What?*: Migrate from Ember to React incrementally by incrementally replacing Ember components with React components.
+
+*Why not?*: The intermediate state of a half-react, half-ember app is something we’d rather not think about -- it’s scary & sad. We’d like to avoid degrading client-side developer experience with this type of complexity. Since this migration will take some time, we feel it more productive to iterate independently.
+
```
Comment (nagarjunakanamarlapudi) [inline docs/rfc/active/react-app/README.md:146]: In the interim ... Any new development happens in react / ember  / both?

- Developing in both can lead to poor developer experience.
- Developing in one can lead to drift.

**Comment by cptran777** [comment]
I think something to note is that an effective incubation of a React app on DataHub should require some modifications and cleanup to the existing Ember code to extract out useful shared modules that are re-usable between the two. Haven't gotten a chance to actually document this yet, but we can have a chat at some point if you'd like to append this RFC with some additional FE/UI information

**Comment by shirshanka** [comment]
> I think something to note is that an effective incubation of a React app on DataHub should require some modifications and cleanup to the existing Ember code to extract out useful shared modules that are re-usable between the two. Haven't gotten a chance to actually document this yet, but we can have a chat at some point if you'd like to append this RFC with some additional FE/UI information

Yes this makes sense. I think the code leverage opportunities will emerge more organically once we start iterating on the React app, thanks for offering to guide us on that @cptran777!

File: docs/rfc/active/react-app/README.md (hunk 2 above)
Comment (shirshanka) [reply docs/rfc/active/react-app/README.md:146]: There are no great answers, once you commit to a dual world for a while. 

Here is a strawman that optimizes for autonomous parallel development: 
1) Get to React-Ember functional parity 
2) Recommend migration from Ember -> React
3) Ember feature freeze 
4) [eventually] deprecate Ember

**Comment by jjoyce0510** [comment]
+1! I've called that out explicitly in the FAQs. Agree we should try to share as much as possible w.r.t. UI components!

**Review by shirshanka — APPROVED** [review]
Thanks for addressing the comments.

File: docs/rfc/active/react-app/README.md (hunk 2 above)
Comment (shirshanka) [reply docs/rfc/active/react-app/README.md:146]: IMO this strategy (what flavor of dual development is sustainable) will emerge over time as well. We'll not lose sight of this as we iterate. Thanks @nagarjunakanamarlapudi!

**Comment by igbopie** [comment]
* What is the proposed architecture? We all agree on those principles, the devil is in the details.
* What UI framework are you going to use?
* How is the react scalable easy to override/config?
* How are you going to maintain code quality standards?
* What if we don't have the time to migrate some components to react? what is the plan around that?

**Pull request merged**